AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +615,877 @@ class AzureIoTHub {
-        // ------------------------------------------------------------------------------------
+            local props = "";
+            if (typeof msg != "message") {
+                throw "Message should be an instance of AzureIoTHub.Message";
+            }
+            if (msg.getProperties() != null) {
+                try {
+                    props = http.urlencode(msg.getProperties());
+                } catch (e) {
+                    _logError("Exception at parsing the properties: " + e);
+                    onSend && onSend(msg, AZURE_IOT_CLIENT_ERROR_GENERAL);
+                    return;
+                }
+            }
+            local topic = _topics.msgSend + props;
+            local reqId = _reqNum;
+            _reqNum++;
 
-        // Open Auth/Event Session/Sender/Receivers
-        // ------------------------------------------------------------------------------------
+            local mqttMsg = _mqttclient.createmessage(topic, msg.getBody(), _options.qos);
 
-        function _openAuthSession() {
-            _sessions.auth <- _connection.opensession(_amqpAuthSessionStatusHandler.bindenv(this));
-        }
+            local msgSentCb = function (err) {
+                if (reqId in _msgPendingQueue) {
+                    delete _msgPendingQueue[reqId];
+                    // TODO: What should we pass to onSend?
+                    onSend && onSend(msg, err);
+                }
+            }.bindenv(this);
 
-        function _openAuthReceiver() {
-            _receivers.auth <- _sessions.auth.openreceiver(AZURE_CLAIM_BASED_SECURITY_PATH, _amqpAuthReceiverStatusHandler.bindenv(this), _authReceiverCB.bindenv(this));
-        }
+            _msgPendingQueue[reqId] <- [msg, onSend];
+            mqttMsg.sendasync(msgSentCb);
+        }
+
+        // Enables or disables message receiving from Azure IoT Hub (https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#receiving-cloud-to-device-messages).
+        //
+        // Parameters:
+        //     onReceive : Function         Callback called every time a new message is received. null disables the feature.
+        //                                  The callback signature:
+        //                                  onReceive(message), where
+        //                                      message :           Received message.
+        //                                          AzureIoTHub.Message
+        //     onDone : Function            Callback called when the operation is completed or an error happens.
+        //          (optional)              The callback signature:
+        //                                  onDone(error), where
+        //                                      error : Integer     0 if the operation is completed successfully, an error code otherwise.
+        //
+        // Returns:                         Nothing.
+        function enableMessageReceiving(onReceive, onDone = null) {
```

Comment:
wouldn't ->`enableIncomingMessages` be better?

Reply:
Done